File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def dfs(graph, v, visited, sub_tree):
    sub_tree[v] = 1
    visited[v] = True
    
    for neighbor in graph[v]:
        if not visited[neighbor]:
            dfs(graph, neighbor, visited, sub_tree)
            sub_tree[v] += sub_tree[neighbor]


def solution(n, wires):
    graph = [[] for _ in range(n+1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    sub_tree = [0] * (n+1)
    visited = [False] * (n+1)
    dfs(graph, 1, visited, sub_tree)
    
    for i in range(1, n+1):
        sub_tree[i] = abs(sub_tree[i] - (n - sub_tree[i]))
    
    return min(sub_tree[1:])

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int minDiff = Integer.MAX_VALUE;
        
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0; i < n+1; i++){
            graph.add(new ArrayList<>()); // 0은 제외 n까지 노드
        }
        
        for(int[] wire : wires){
            int a = wire[0];
            int b = wire[1];
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        boolean[] visited = new boolean[n+1];
        int[] subTree = new int[n+1];
        dfs(graph, visited, 1, subTree);
       
        for(int i = 0; i < subTree.length; i++){
            subTree[i] = Math.abs(subTree[i] - (n - subTree[i]));
        }
        
        for(int i = 2; i < subTree.length; i++){
            minDiff = Math.min(minDiff, subTree[i]);
        }
        return minDiff;
    }
    
    private void dfs(List<List<Integer>> graph, boolean[] visited, int node, int[] subTree){
        visited[node] = true;
        subTree[node] = 1; // 자기 자신
        
        for(int next : graph.get(node)){
            if(!visited[next]){
                dfs(graph, visited, next, subTree);
                subTree[node] += subTree[next];
            }
        }
        
    }
}